src/main/java/uk/gov/companieshouse/itemhandler/service/FilingHistoryDescriptionProviderService.java

```diff
@@ -0,0 +1,121 @@
+        Map<String, String> filingHistoryDescriptionsLocal = null;
+        try(final InputStream inputStream = new FileInputStream(filingHistoryDescriptionsFile)) {
+            final Yaml yaml = new Yaml();
+            final Map<String, Object> filingHistoryDescriptionsRoot = yaml.load(inputStream);
+            filingHistoryDescriptionsLocal =
+                    (Map<String, String>) filingHistoryDescriptionsRoot.get(FILING_HISTORY_DESCRIPTION_KEY);
+            if (filingHistoryDescriptionsLocal == null) {
+                Map<String, Object> logMap = new HashMap<>();
+                logMap.put(DESCRIPTION_LOG_KEY, FILING_HISTORY_DESCRIPTION_KEY);
+                LoggingUtils.getLogger().error("Filing History descriptions file not found", logMap);
+                return null;
+            }
+        } catch (IOException ioe) {
+            // This is very unlikely to happen here given File.exists() check above,
+            // and that it is not likely to encounter an error closing the stream either.
+        }
+        return filingHistoryDescriptionsLocal;
+    }
+
+    private String getFilingHistoryDescriptionWithKey(String descriptionKey) {
+        if(descriptionKey == null || descriptionKey.isEmpty()) {
+            return descriptionKey;
+        } else {
+            return filingHistoryDescriptions.get(descriptionKey);
+        }
+    }
+
+    /**
+     * Fetches the specified key from the filing history enumerations and replaces the variables
+     * with the ones in the description values
+     * @param descriptionKey the key to fetch from the filing history enumerations
+     * @param filingHistoryDescriptionValues the map of the description values to replace in the description
+     * @return the formatted filing history description
+     */
+    public String mapFilingHistoryDescription(String descriptionKey, Map<String, Object> filingHistoryDescriptionValues) {
+        if(filingHistoryDescriptions == null) {
+            return null;
+        }
+        String description = getFilingHistoryDescriptionWithKey(descriptionKey);
+        if(filingHistoryDescriptionValues.containsKey("description")) {
+            return (String) filingHistoryDescriptionValues.get("description");
+        } else {
+            final StringBuilder sb = new StringBuilder(description);
+            replaceAll(sb, "*", ""); // remove all asterisks in fetched string
+            filingHistoryDescriptionValues.forEach((k,v)-> {
+                String value = k.contains("date") ? reformatDateFiled((String)v) : (String) v;
+                replaceAll(sb, "{"+k+"}", value);
+            });
+            return sb.toString();
+        }
+    }
+
+    /**
+     * Reformats a date filed string such as "2009-08-23" as "23 August 2009".
+     * @param dateFiled the date filed as reported from the filing history
+     * @return the same date rendered for display purposes
+     */
+    public static String reformatDateFiled(final String dateFiled) {
```

Comment:
Why do we now have two `reformatDateFiled` functions?

Reply:
I can rename this. This is slightly different to the other one. This one does "2009-08-23" -> "23 August 2009". Where as the other one does "2009-08-23" -> "23 Aug 2009".

Reply:
OK I see. Ideally all these similar but slightly different formats would be declared as constants in the same place for ease of understanding.

Reply:
Yeh so there is a OrderDataToOrderConfirmationMapperConstants file. But because this file is a service. It doesn't quite fit into it. Shall I rename OrderDataToOrderConfirmationMapperConstants to DateConstants and move it to a util folder?

Reply:
Sounds good.

Reply:
Actually, I think I have another question. Is the date in question here actually the `date filed` or some kind of `action(ed)` date?

Reply:
It is the date the filing history event occured, which can be different from the date filed. e.g. 

23 Jun 2009 | Group of companies' accounts made up to 31 August 2008

Reply:
So the names we use need to reflect this distinction.

Reply:
The date on the left (23 Jun 2009) is the date filed. Whereas the other date is the date which the event occured, one which is in the filing history description.

Reply:
Yes so perhaps I can call the method reformatFilingHistoryDescriptionDate?

Reply:
Why not use the term `dateOccurred` if that is what this date really represents? I see there is a confusing number of different dates when we get filing history items from the API, for example:

```
        {
            "date": "2015-02-02",
            "action_date": "2014-08-31",
            "description_values": {
                "made_up_date": "2014-08-31"
            },
            "type": "AA",
            "category": "accounts",
            "links": {
                "self": "/company/00006400/filing-history/MzExNjA3NjM0MmFkaXF6a2N4"
            },
            "description": "accounts-with-accounts-type-group",
            "paper_filed": true,
            "barcode": "A3ZNU6MJ",
            "transaction_id": "MzExNjA3NjM0MmFkaXF6a2N4"
        },
```

Do you know which of the above dates this is?

Reply:
Perhaps `made_up_date` in this case, which seems to be the same as the `action_date`?

Reply:
Yeh I can change it to dateOccured. So the app does a look up in the filing history enumerations file for accounts-with-accounts-type-group. This will bring back **Group of companies' accounts** made up to {made_up_date}. The service will replace {made_up_date} with a formatted version of made_up_date in description_values.

Reply:
OK. Looking at the result of a query to the filing history API, it seems to be that in general, where there is some kind of `...date` in the `description_values`, the same value seems to be present as the `action_date` too. So perhaps `action_date` is the generic name for these dates that has already been chosen within CH? You could check this with @macreman perhaps?

Reply:
That's not always the case, here is one from my local database
```
{
    "_id" : "MjAzNTYyNTE5M2FkaXF6a2N4",
    "data" : {
        "action_date" : ISODate("2008-08-31T00:00:00.000Z"),
        "category" : "accounts",
        "date" : ISODate("2009-06-23T11:04:31.000Z"),
        "description" : "accounts-with-accounts-type-group",
        "type" : "AA",
        "description_values" : {
            "made_up_date" : ISODate("2008-08-31T00:00:00.000Z")
        },
        "paper_filed" : true,
        "links" : {
            "document_metadata" : "/document/WLlqKpXH7axcPda4xkIE4v-AxyXkzhY5NyWwVKcTT9w",
            "self" : "/company/00006400/filing-history/MjAzNTYyNTE5M2FkaXF6a2N4"
        },
        "pages" : 32
    },
    "_document_id" : "000L4O9HAXA3787",
    "_entity_id" : "2035625193",
    "company_number" : "00006400",
    "_barcode" : "L4O9HAXA",
    "original_description" : "Group of companies' accounts made up to 31/08/08"
}
```